packages/geo/src/lib/datasource/shared/datasources/wfs.service.ts

```diff
@@ -54,96 +55,31 @@ export class WFSService extends DataService {
-    }
-    return Object.assign({}, wfsDataSourceOptions, {
-      wfsCapabilities: { xmlBody: '', GetPropertyValue: false }
-    });
-  }
-
-  public buildBaseWfsUrl(
-    wfsDataSourceOptions: WFSDataSourceOptions,
-    wfsQuery: string
-  ): string {
-    let paramTypename = 'typename';
-    if (
-      wfsDataSourceOptions.paramsWFS.version === '2.0.0' ||
-      !wfsDataSourceOptions.paramsWFS.version
-    ) {
-      paramTypename = 'typenames';
-    }
-    const baseWfsQuery = 'service=wfs&request=' + wfsQuery;
-    const wfsTypeName =
-      paramTypename + '=' + wfsDataSourceOptions.paramsWFS.featureTypes;
-    const wfsVersion = wfsDataSourceOptions.paramsWFS.version
-      ? 'version=' + wfsDataSourceOptions.paramsWFS.version
-      : 'version=' + '2.0.0';
-
-    return `${
-      wfsDataSourceOptions.urlWfs
-    }?${baseWfsQuery}&${wfsVersion}&${wfsTypeName}`;
-  }
-
-  public wfsGetFeature(
+  private wfsGetFeature(
     wfsDataSourceOptions: WFSDataSourceOptions,
-    nb = 5000,
-    epsgCode = 3857,
-    propertyname = ''
+    nb = defaultMaxFeatures,
+    epsgCode = defaultEpsg,
+    propertyname?
   ): Observable<any> {
-    const baseUrl = this.buildBaseWfsUrl(wfsDataSourceOptions, 'GetFeature');
-    const outputFormat = wfsDataSourceOptions.paramsWFS.outputFormat
-      ? 'outputFormat=' + wfsDataSourceOptions.paramsWFS.outputFormat
-      : '';
-    const srsname = wfsDataSourceOptions.paramsWFS.srsName
-      ? 'srsname=' + wfsDataSourceOptions.paramsWFS.srsName
-      : 'srsname=EPSG:' + epsgCode;
-    const wfspropertyname =
-      propertyname === '' ? propertyname : '&propertyname=' + propertyname;
-    let paramMaxFeatures = 'maxFeatures';
-    if (
-      wfsDataSourceOptions.paramsWFS.version === '2.0.0' ||
-      !wfsDataSourceOptions.paramsWFS.version
-    ) {
-      paramMaxFeatures = 'count';
-    }
-
-    let maxFeatures;
-    if (nb !== 5000) {
-      maxFeatures = paramMaxFeatures + '=' + nb;
+    const queryStringValues = formatWFSQueryString(wfsDataSourceOptions, nb, epsgCode, propertyname);
+    const baseUrl = queryStringValues.find(f => f.name === 'getfeature').value;
+    const outputFormat = wfsDataSourceOptions.paramsWFS.outputFormat;
+    if (gmlRegex.test(outputFormat) || !outputFormat) {
+      return this.http.get(baseUrl, { responseType: 'text' });
     } else {
-      maxFeatures = wfsDataSourceOptions.paramsWFS.maxFeatures
-        ? paramMaxFeatures + '=' + wfsDataSourceOptions.paramsWFS.maxFeatures
-        : paramMaxFeatures + '=' + nb;
-    }
-    const urlWfs = `${baseUrl}&${outputFormat}&${srsname}&${maxFeatures}${wfspropertyname}`;
-    const patternGml = new RegExp('.*?gml.*?');
-    if (
-      patternGml.test(wfsDataSourceOptions.paramsWFS.outputFormat.toLowerCase())
-    ) {
-      return this.http.get(urlWfs, { responseType: 'text' });
-    } else {
-      return this.http.get(urlWfs);
+      return this.http.get(baseUrl);
     }
   }
 
   public getValueFromWfsGetPropertyValues(
```

Comment:
Have you considered rxjs retry and concatMap operators?

Reply:
No because the if clause is use to detect if openlayers need to get the response as text to parse the gml or in a json structure... I don't think it's an use case for these operators?

Reply:
concatMap simply returns a new Observable. It would probably work here. I haven't used retry a lot but retryWhen would probably do as well. No need to spend a lot of time testing those but I think it's worth testing it.

Reply:
Code has been simplified. This is no longer applicable.